utils/misc.mli

```diff
@@ -486,3 +486,153 @@ module EnvLazy: sig
+
+  type version = int
+
+  type kind =
+    | Exec
+    | Cmi | Cmo | Cma
+    | Cmx of native_obj_config | Cmxa of native_obj_config
+    | Cmxs
+    | Cmt | Ast_impl | Ast_intf
+
+  type raw = string
+  (** the type of raw magic numbers, such as "Caml1999O011" *)
+
+  type info = {
+    kind: kind;
+    version: version;
+    (** Note: some versions of the compiler use the same [version] suffix
+        for all kinds, but others use different versions counters for different
+        kinds. We may only assume that versions are growing monotonically
+        (not necessarily always by one) between compiler versions. *)
+  }
+
+  type parse_error =
+    | Truncated of string
+    | Not_a_magic_number of string
+
+  val explain_parse_error : kind option -> parse_error -> string
+  (** Produces an explanation for a parse error. If no kind is provided,
+      we use an unspecific formulation suggesting that any compiler-produced
+      object file would have been satisfying. *)
+
+  val parse : raw -> (info, parse_error) result
+  (** Parses a raw magic number *)
+
+  val read_info : in_channel -> (info, parse_error) result
+  (** Read a raw magic number from an input channel.
+
+      If the data read [str] is not a valid magic number, it can be
+      recovered from the [Truncated str | Not_a_magic_number str]
+      payload of the [Error parse_error] case.
+
+      If parsing succeeds with an [Ok info] result, we know that
+      exactly [magic_length] bytes have been consumed from the
+      input_channel. *)
+
+  val magic_length : int
+  (** all magic numbers take the same number of bytes *)
+
+  val string_of_kind : kind -> string
+  (** a user-printable string for a kind, eg. "exec" or "cmo", to use
+      in error messages. *)
+
+  val human_name_of_kind : kind -> string
+  (** a user-meaningful name for a kind, eg. "executable file" or
+      "bytecode object file", to use in error messages. *)
+
+  val current_raw : kind -> raw
+  (** the current magic number of each kind *)
+
+  val current_version : kind -> version
+  (** the current version of each kind *)
+
+  type 'a unexpected = { expected : 'a; actual : 'a }
+  type unexpected_error =
+    | Kind of kind unexpected
+    | Version of kind * version unexpected
+
+  val check_current : kind -> info -> (unit, unexpected_error) result
+  (** [check_current kind info] checks that the provided magic [info]
+      is the current version of [kind]'s magic header. *)
+
+  val explain_unexpected_error : unexpected_error -> string
+  (** Provides an explanation of the [unexpected_error]. *)
+
+  type error =
+    | Parse_error of parse_error
+    | Unexpected_error of unexpected_error
+
+  val read_current_info :
+    expected_kind:kind option -> in_channel -> (info, error) result
+  (** Read a magic number, and check that it is the current version as its kind.
+      If the [expected_kind] argument is [None], any kind is accepted. *)
```

Comment:
Why not an optional argument ?

Reply:
Also, the documentation should mention that this behaves as `read_info` wrt the channel.

Reply:
Optional arguments are generally a bad idea for API design, unless we have strong confidence that the default choice is the correct "path of least effort" that will do the right thing for users who don't read the documentation carefully. In this case it's pretty clear that there is no correct default value for this value, and users really need to think about what they should put in this argument.

In fact an intermediary state of this PR had two separate functions, one for reading a "current" number at any possible kind, and one for reading a current number at a specific expected kind. I think that the present approach is better (at least on the implementation side it reduces duplication), but it's important that the choice be explicit.

(Good point on the documentation.)

Reply:
The unspoken assumption that any labelled argument of option type could be made optional is bad and we should get rid of it!

Reply:
I updated the documentation to mention `read_info`.